lib/langchain/tool/news_retriever/news_retriever.rb

```diff
@@ -0,0 +1,132 @@
+    # @param from [String] A date and optional time for the oldest article allowed. This should be in ISO 8601 format.
+    # @param to [String] A date and optional time for the newest article allowed. This should be in ISO 8601 format.
+    # @param language [String] The 2-letter ISO-639-1 code of the language you want to get headlines for. Possible options: ar, de, en, es, fr, he, it, nl, no, pt, ru, se, ud, zh.
+    # @param sort_by [String] The order to sort the articles in. Possible options: relevancy, popularity, publishedAt.
+    # @param page_size [Integer] The number of results to return per page. 20 is the API's default, 100 is the maximum. Our default is 5.
+    # @param page [Integer] Use this to page through the results.
+    #
+    # @return [String] JSON response
+    def get_everything(
+      q: nil,
+      search_in: nil,
+      sources: nil,
+      domains: nil,
+      exclude_domains: nil,
+      from: nil,
+      to: nil,
+      language: nil,
+      sort_by: nil,
+      page_size: 5, # The API default is 20 but that's too many.
+      page: nil
+    )
+      Langchain.logger.info("Retrieving all news", for: self.class)
+
+      params = {apiKey: @api_key}
+      params[:q] = q if q
+      params[:searchIn] = search_in if search_in
+      params[:sources] = sources if sources
+      params[:domains] = domains if domains
+      params[:excludeDomains] = exclude_domains if exclude_domains
+      params[:from] = from if from
+      params[:to] = to if to
+      params[:language] = language if language
+      params[:sortBy] = sort_by if sort_by
+      params[:pageSize] = page_size if page_size
+      params[:page] = page if page
+
+      send_request(path: "everything", params: params)
+    end
+
+    # Retrieve top headlines
+    #
+    # @param country [String] The 2-letter ISO 3166-1 code of the country you want to get headlines for. Possible options: ae, ar, at, au, be, bg, br, ca, ch, cn, co, cu, cz, de, eg, fr, gb, gr, hk, hu, id, ie, il, in, it, jp, kr, lt, lv, ma, mx, my, ng, nl, no, nz, ph, pl, pt, ro, rs, ru, sa, se, sg, si, sk, th, tr, tw, ua, us, ve, za.
+    # @param category [String] The category you want to get headlines for. Possible options: business, entertainment, general, health, science, sports, technology.
+    # @param sources [String] A comma-seperated string of identifiers for the news sources or blogs you want headlines from. Use the /sources endpoint to locate these programmatically.
+    # @param q [String] Keywords or a phrase to search for.
+    # @param page_size [Integer] The number of results to return per page. 20 is the API's default, 100 is the maximum. Our default is 5.
+    # @param page [Integer] Use this to page through the results.
+    #
+    # @return [String] JSON response
+    def get_top_headlines(
+      country: nil,
+      category: nil,
+      sources: nil,
+      q: nil,
```

Comment:
You probably know, but missing some parameters like 'language', 'from', 'to',...
So currently if you ask, for example, for "Top headlines from Germany in English" the tool calls fails with `unknown keyword: :language (ArgumentError)` and leaves the Assistant in a bad state (missing role=tool messages) and every subsequent attempt fails like below (with OpenAI):
```
3.2.1 :014 > assistant.add_message_and_run content:"Now give the news in Germany?", auto_tool_execution: true
I, [2024-05-06T20:49:45.668612 #8222]  INFO -- : [Langchain.rb] [Langchain::Assistant]: Sending a call to Langchain::LLM::OpenAI
OpenAI HTTP Error (spotted in ruby-openai 6.4.0): {"error"=>{"message"=>"An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'. The following tool_call_ids did not have response messages: call_u4qPhNeJInxgjIgAQQ32AuE2", "type"=>"invalid_request_error", "param"=>"messages.[3].role", "code"=>nil}}
```

Reply:
I added `language` but unfortunately it doesn't seem to like translating headlines to English:
"I couldn't find any latest news from Germany in English. Would you like me to search for news from a different country or on a different topic?"

I also saw OpenAI sometimes not pass the ':language' parameter to the tool and try multiple times since the headlines returned weren't English.  Wierd.

Reply:
Actually your code looks fine. Sorry. I think you have all of the correct parameters.  

Although it does seem the llm tries to pass `language` to top_headlines even though it's not listed in allowed params.  Try it: "Top headlines from Germany in English".